api/python/cellxgene_census/src/cellxgene_census/experimental/ml/pytorch.py

```diff
@@ -131,34 +131,33 @@ def __next__(self) -> ObsDatum:
         while len(obs) < self.batch_size:
             try:
                 obs_partial, X_partial = self._read_partial_torch_batch(self.batch_size)
-                if X is None:
-                    obs = obs_partial
-                    X = X_partial
-                else:
-                    obs = pd.concat([obs, obs_partial], axis=0)
-                    X = sparse.vstack([X, X_partial])
-
+                obs = pd.concat([obs, obs_partial], axis=0)
+                X = sparse.vstack([X, X_partial])
             except StopIteration:
                 break
 
         if len(obs) == 0:
             raise StopIteration
 
+        assert obs["soma_joinid"].is_unique
+
         obs_encoded = pd.DataFrame(
             data={"soma_joinid": obs.soma_joinid}, columns=obs.columns, dtype=np.int32, index=obs.index
         )
         for col, enc in self.encoders.items():
             obs_encoded[col] = enc.transform(obs[col])
 
-        obs_tensor = torch.Tensor(obs_encoded.to_numpy()).int()
+        # `as_tensor()` avoids copying the numpy array data
+        obs_tensor = torch.from_numpy(obs_encoded.to_numpy(dtype=np.int32))
+
         if not self.sparse_X:
-            X_tensor = torch.Tensor(X.todense())
+            X_tensor = torch.from_numpy(X.todense())
```

Comment:
Not related to bug fix. But use same tensor constructor as above, for memory usage improvement.